balanced-optiml/bal-optiml-library/library.whizzml

```diff
@@ -0,0 +1,139 @@
+                             " Only categorical fields can be used.")
+              "code" 101}))
+    (when (not (member? rare-class (field-categories field)))
+      (raise {"message" (str "The objective field "
+                             objective-field
+                             " does not contain the "
+                             rare-class
+                             " category. Please check its name.")
+              "code" 102}))))
+
+;; Function that splits a dataset in training, validation and test datasets
+(define (_create-train-validation-test ds-id train-rate seed)
+  (let (first-split (create-dataset-split ds-id train-rate seed)
+        second-split (create-dataset-split (last first-split) 0.5 seed))
+    (cons (head first-split) second-split)))
+
+
+;; Function to apply stratified split for the rare class of the objective field
+;; with a given train
+(define (rare-stratified-split-fn ds-id training-rate objective-field
+                                  rare-class project)
+  (let (dataset (fetch ds-id)
+        ds-name (dataset "name")
+        fields (resource-fields dataset)
+        _ (_check-cat fields objective-field rare-class)
+        objective-id ((find-field fields objective-field) "id")
+        filters [(flatline "(= (f {{objective-id}}) {{rare-class}})")
+                 (flatline "(!= (f {{objective-id}}) {{rare-class}})")]
+        names [rare-class (str "not " rare-class)]
+        cat-ds-ids (iterate (acc [] expr filters name names)
+                     (append acc
+                       (_create-train-validation-test
+                         (create-dataset {"origin_dataset" ds-id
+                                          "lisp_filter" expr
+                                          "name" (str (dataset "name")
+                                                      " - "
+                                                      name)
+                                          "project" project})
+                         training-rate
+                         "bigml")))
+        inner-ds (flatten cat-ds-ids)
+        ds-ids (map (lambda (ids name)
+                      (merge-datasets ids {"name" name}))
+                    (_transpose cat-ds-ids)
+                    [(str ds-name " - train")
+                     (str ds-name " - validation")
+                     (str ds-name " - test")]))
+    (wait* ds-ids)
+    (map _safe-delete inner-ds)
+    ds-ids))
+
+
+;; Function to generated datasets that keep all instances of the rare class
+;; while undersampling the instances in the common class to maintain a
+;; proportion with the rare class
+(define (undersample-fn ds-id number-of-samples objective-field rare-class
+         proportion)
+  (let (dataset (fetch ds-id)
+        ds-name (dataset "name")
+        fields (resource-fields dataset)
+        _ (_check-cat fields objective-field rare-class)
+        objective-id ((find-field fields objective-field) "id")
```

Comment:
another optional nit: you could have a function _find-objective that performs the check and save a tiny bit of repetition.